fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,814 @@
+                if term_val == "xterm-256color" {
+                    const SITM_ESC: &str = "\x1B[3m";
+                    const RITM_ESC: &str = "\x1B[23m";
+                    const DIM_ESC: &str = "\x1B[2m";
+
+                    if term.get(curses::ENTER_ITALICS_MODE).is_none() {
+                        term.set(curses::ENTER_ITALICS_MODE, Some(SITM_ESC));
+                    }
+                    if term.get(curses::EXIT_ITALICS_MODE).is_none() {
+                        term.set(curses::EXIT_ITALICS_MODE, Some(RITM_ESC));
+                    }
+                    if term.get(curses::ENTER_DIM_MODE).is_none() {
+                        term.set(curses::ENTER_DIM_MODE, Some(DIM_ESC));
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// This is a pretty lame heuristic for detecting terminals that do not support setting the title.
+/// If we recognise the terminal name as that of a virtual terminal, we assume it supports setting
+/// the title. If we recognise it as that of a console, we assume it does not support setting the
+/// title. Otherwise we check the ttyname and see if we believe it is a virtual terminal.
+///
+/// One situation in which this breaks down is with screen, since screen supports setting the
+/// terminal title if the underlying terminal does so, but will print garbage on terminals that
+/// don't. Since we can't see the underlying terminal below screen there is no way to fix this.
+fn does_term_support_setting_title(vars: &dyn Environment) -> bool {
+    #[rustfmt::skip]
+    const TITLE_TERMS: &[&wstr] = &[
+        L!("xterm"), L!("screen"),    L!("tmux"),    L!("nxterm"),
+        L!("rxvt"),  L!("alacritty"), L!("wezterm"),
+    ];
+
+    let Some(term) = vars.get_unless_empty(L!("TERM")).map(|v| v.as_string()) else {
+        return false;
+    };
+    let term: &wstr = term.as_ref();
+
+    let recognized = TITLE_TERMS.contains(&term)
+        || term.starts_with(L!("xterm-"))
+        || term.starts_with(L!("screen-"))
+        || term.starts_with(L!("tmux-"));
+    if !recognized {
+        if [
+            L!("linux"),
+            L!("dumb"),
+            L!("vt100"), // NetBSD
+            L!("wsvt25"),
+        ]
+        .contains(&term)
+        {
+            return false;
+        }
+
+        let mut buf = [b'\0'; libc::PATH_MAX as usize];
+        let retval =
+            unsafe { libc::ttyname_r(libc::STDIN_FILENO, buf.as_mut_ptr().cast(), buf.len()) };
+        let buf = &buf[..buf.iter().position(|c| *c == b'\0').unwrap()];
+        if retval != 0
+            || buf.windows(b"tty".len()).any(|w| w == b"tty")
+            || buf.windows(b"/vc/".len()).any(|w| w == b"/vc/")
+        {
+            return false;
+        }
+    }
+
+    true
+}
+
+// Initialize the curses subsystem
+fn init_curses(vars: &dyn Environment) {
+    for var_name in CURSES_VARIABLES {
+        let name = var_name.to_string();
```

Comment:
This is where `String::ToCString` gets used, but we should not introduce that because then we are adding APIs which fundamentally do the wrong thing, and future contributors will be tempted to use them.

`CURSES_VARIABLES` should contain `&wstr` or a byte string literal. We should not introduce types that assume UTF8 encoding like `str` or `String`.

Reply:
See my not above. I'll change the impl to convert to a CString safely, respecting the user's locale.